File: services/product-service/internal/domain/plan.go
package domain

import (
	"context"
	"time"
)

type Plan struct {
	ID                 int64
	Name               string
	Slug               string
	IntervalID         int64
	ProductID          int64
	IsActive           bool
	HasTrial           bool
	TrialIntervalID    *int64 // nullable
	IntervalCount      int32
	TrialIntervalCount int32
	Description        string
	Type               string // flat_rate, per_unit, tiered
	MaxUsersPerTenant  int32
	MeterID            *int64 // nullable
	IsVisible          bool
	CreatedAt          time.Time
	UpdatedAt          time.Time
}

type PlanRepository interface {
	GetByID(ctx context.Context, id int64) (*Plan, error)
	GetBySlug(ctx context.Context, slug string) (*Plan, error)
	GetByProduct(ctx context.Context, productID int64) ([]*Plan, error)
	Create(ctx context.Context, plan *Plan) error
	Update(ctx context.Context, plan *Plan) error
	Delete(ctx context.Context, id int64) error
	GetAll(ctx context.Context, page, perPage int, activeOnly, visibleOnly bool) ([]*Plan, int, error)
}

type PlanService interface {
	GetByID(ctx context.Context, id int64) (*Plan, error)
	GetBySlug(ctx context.Context, slug string) (*Plan, error)
	GetByProduct(ctx context.Context, productID int64) ([]*Plan, error)
	Create(ctx context.Context, plan *Plan) error
	Update(ctx context.Context, plan *Plan) error
	Delete(ctx context.Context, id int64) error
	GetAll(ctx context.Context, page, perPage int, activeOnly, visibleOnly bool) ([]*Plan, int, error)
}

File: web/src/types/plan.ts
export interface Plan {
  id: string
  name: string
  slug: string
  intervalId: string
  productId: string
  isActive: boolean
  hasTrial: boolean
  trialIntervalId?: string | null
  intervalCount: number
  trialIntervalCount?: number | null
  description?: string | null
  type: string
  maxUsersPerTenant?: number | null
  meterId?: string | null
  isVisible: boolean
  createdAt: number
  updatedAt: number
}

export interface PlansResponse {
  plans: {
    success: boolean
    message: string
    data: {
      plans: Plan[]
      total: number
      page: number
      perPage: number
    }
  }
}

export interface PlanResponse {
  plan: {
    success: boolean
    message: string
    data: Plan
  }
}

export interface CreatePlanInput {
  name: string
  slug?: string
  intervalId: string
  productId: string
  intervalCount: number
  type: string
  isActive?: boolean
  hasTrial?: boolean
  trialIntervalId?: string
  trialIntervalCount?: number
  description?: string
  maxUsersPerTenant?: number
  meterId?: string
  isVisible?: boolean
}

export interface UpdatePlanInput {
  name: string
  slug?: string
  intervalId: string
  intervalCount: number
  type: string
  isActive?: boolean
  hasTrial?: boolean
  trialIntervalId?: string
  trialIntervalCount?: number
  description?: string
  maxUsersPerTenant?: number
  meterId?: string
  isVisible?: boolean
}

export interface CreatePlanResponse {
  createPlan: {
    success: boolean
    message: string
    data: Plan
  }
}

export interface UpdatePlanResponse {
  updatePlan: {
    success: boolean
    message: string
    data: Plan
  }
}

export interface DeletePlanResponse {
  deletePlan: {
    success: boolean
    message: string
  }
}
